### ISSUE TITLE
Do dynamic imports still make sense?

### ISSUE THREAD

**Issue by mercmobily** [open-wc/open-wc#1551]
I have a pretty big application coming from the old Polymer Starter Kit. They were very big on the PRPL pattern. One of the main concept was to load parts only if needed.
This is a bit of a philosophical question, but... does it actually still make sense to do that? At the moment in the built environment I have one bigg-ish file of 133185kb -- the rest is a huge bunch of small-ish files, totalling around 200kb.

However, loading on demand has its problems: the logic to re-load if anything goes wrong is difficult; giving "loading" feedback to users is also difficult. Plus, users might not want to wait a couple of seconds to see the next page.

Do you recommend using dynamic imports in an app? Or would you recommend having one bigger file with everything in it?

Basically: have dynamic imports fallen out of best practices? What's your take on them?

**Comment by LarsDenBakker** [comment]
Dynamic imports still definitely make sense, you'd have have to measure and evaluate for each application. If have a SPA and it is made out of pages, you're going to get the most benefit from lazy loading each page on navigation. Code would also be broken up like this for a regular website.

You can go further by lazy loading more components on user interaction. if you're concerned about loading states you can preload things on the background, but then at least the initial payload is smaller so that you can start showing things to the user as soon as possible.

**Comment by bennypowers** [comment]
An alternative to lazy loading is a multipage architecture, perhaps coupled with a templating service worker

**Comment by mercmobily** [comment]
@LarsDenBakker  The idea of loading the home page on load time, and then in the background loading the rest of the app, definitely makes sense. Doing so would be quite trivial on a superficial level: you could just load the main app, and then 5 seconds later load the rest of the components with a simple `import()`. But, the devil is in the detail: what about loading failures? The background loader would need to check the result of each load, and at least retry if it fails. This would imply having a hash of pages to be loaded, and then attempting to import them in the background -- and re-attempting the import if things go wrong. This would have to run each time, at -- say -- 5 seconds interval. This would need to happen even if the network is down, with the assumption that the service worker will provide those files.
I mean... it gets pretty complicated pretty quickly, doesn't it?

**Comment by LarsDenBakker** [comment]
Indeed, things can get complicated. This is a tradeoff to make for your application, we don't really have any documentation for this from open-wc. Perhaps this is more something for https://web.dev/ ? Going to to close this one to keep the issue focused on bugs/feature requests.
